src/core/memory.d

```diff
@@ -919,3 +919,186 @@ extern (C) private pure @system @nogc nothrow
+
+In detail, `__delete` does the following:
+$(UL
+
+    $(LI If `t` is a class object reference, and
+        there is a destructor for that class, the destructor
+        is called for that object instance.
+    )
+    $(LI Otherwise, the garbage collector is called to immediately free the
+        memory allocated for the class instance or struct instance.
+        If the garbage collector was not used to allocate the memory for
+        the instance, undefined behavior will result.
+    )
+    $(LI If `t` is a pointer or a dynamic array,
+        the garbage collector is called to immediately release the
+        memory.
+        If the garbage collector was not used to allocate the memory for
+        the instance, undefined behavior will result.
+    )
+    $(LI The pointer, dynamic array, or reference is set to `null`
+        after the delete is performed.
+        Any attempt to reference the data after the deletion via another
+        reference to it will result in undefined behavior.
+    )
+)
+
+Params:
+    t = aggregate object that should be destroyed
+
+See_Also: $(REF destroy, object), $(REF free, core,GC)
+*/
+void __delete(T)(ref T t) @system
+{
+    static if (is(T == interface))
+    {
+        _d_delinterface(cast(void**) &t);
+    }
+    else static if (is(T == class))
+    {
+        _d_delclass(cast(Object*) &t);
+    }
+    else static if (is(T == struct))
+    {
+        _d_delstruct(cast(void**) &t, typeid(T));
+    }
+    else static if (is(T == U*, U))
+    {
+        t = null;
+    }
+    else static if (is(T : E[], E))
+    {
+        t = null;
+    }
+    else
+    {
+        static assert(0, "It is not possible to delete: " ~ T.stringof);
+    }
+}
+
+/// Deleting interfaces
+unittest
+{
+    bool dtorCalled;
+    class B
+    {
+        int test;
+        ~this()
+        {
+            dtorCalled = true;
+        }
+    }
+    B b = new B();
+    B a = b;
+    b.test = 10;
+
+    __delete(b);
+    assert(b is null);
+    assert(dtorCalled);
+    // but a still points to it
+    assert(a !is null);
+
+}
+
+/// Deleting classes
+unittest
+{
+    class B
+    {
+        string a;
+    }
+    B b = new B();
+    b.a = "exists";
+    auto s = b.a;
+
+    __delete(b);
+    assert(b is null);
+    assert(s == "exists");
+}
+
+/// Deleting interfaces
+unittest
+{
+    interface A
+    {
+        int quack();
+    }
+    class B : A
+    {
+        int a;
+        int quack()
+        {
+            a++;
+            return a;
+        }
+    }
+    A a = new B();
+    a.quack();
+
+    __delete(a);
+    assert(a is null);
+}
+
+/// Deleting structs
+unittest
+{
+    struct A
+    {
+        string test;
+    }
+    A a = {
+        test: "exists",
+    };
+    auto b = &a;
+    __delete(b);
+    assert(a.test == "exists");
+
+    __delete(a);
+    assert(a == A.init);
+}
+
+/// Deleting arrays
+unittest
+{
+    int[] a = [1, 2, 3];
+    auto b = a;
+    __delete(b);
+    assert(b is null);
+    assert(a == [1, 2, 3]);
```

Comment:
If I'm understanding correctly, you are access through `a` after `b` was `__delete`ed.  According to the documentation comments, that's undefined behavior.